File: Week01/d6_matrix_transpose.c
/**
 * Author: Penut Chen
 * Date: 2019-7-1
 * Description: 
 *   Practice of matrix transpose.
 */

#include <stdio.h>
#include <stdlib.h>

int** generate_matrix(int, int);
void display_matrix(int**, int, int);
int** matrix_transpose(int**, int, int);

int main() {
    int row = 3, col = 5;
    int **a = generate_matrix(row, col);
    
    printf("Before Transpose\n");
    display_matrix(a, row, col);

    int **b = matrix_transpose(a, row, col);
    
    printf("\nAfter Transpose\n");
    display_matrix(b, col, row);
    
    free(a);
    free(b);
    
    return 0;
}

int** generate_matrix(int row, int col) {
    int **matrix = (int**)malloc(sizeof(int*) * row);
    for (int i = 0; i < row; i++) {
        matrix[i] = (int*)malloc(sizeof(int) * col);
        for (int j = 0; j < col; j++)
            matrix[i][j] = rand() % 90 + 10;
    }
    return matrix;
}

void display_matrix(int **matrix, int row, int col) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++)
            printf("%3d ", matrix[i][j]);
        printf("\n");
    }
}

int** matrix_transpose(int **matrix, int row, int col) {
    int nrow = col, ncol = row;
    int **nmatrix = (int**)malloc(sizeof(int) * nrow);

    for (int i = 0; i < nrow; i++) {
        nmatrix[i] = (int*)malloc(sizeof(int) * ncol);
        for (int j = 0; j < ncol; j++)
            nmatrix[i][j] = matrix[j][i];
    }

    return nmatrix;
}

File: Week01/d6_matrix_transpose.py
"""
Author: Penut Chen
Date: 2019-7-1
Description:
  Practice of matrix transpose.
"""

import random

def main():
    a = Matrix(size=(3, 5), name='Matrix A')

    print(a, end='\n\n')
    print(a.transpose())

class Matrix:
    def __init__(self, matrix=None, size=None, name='matrix'):
        self.size = size
        self.row, self.col = size
        self.name = name
        if matrix is None:
            self.matrix = [[random.randint(10, 99) for i in range(self.col)] for j in range(self.row)]
        else:
            self.matrix = [[j for j in i] for i in matrix]

    def __str__(self):
        m = '\n'.join([' '.join(['%3d' % x for x in row]) for row in self.matrix])
        return '%s\n%s' % (self.name, m)

    def __add__(self, another):
        row, col = self.row, self.col
        m = [[self.matrix[i][j] + another.matrix[i][j] for j in range(col)] for i in range(row)]
        return Matrix(matrix=m, size=self.size, name='%s + %s' % (self.name, another.name))
    
    def transpose(self):
        row, col = self.col, self.row
        m = [[self.matrix[j][i] for j in range(col)] for i in range(row)]
        return Matrix(matrix=m, size=(row, col), name='Transpose of %s' % self.name)

if __name__ == "__main__":
    main()